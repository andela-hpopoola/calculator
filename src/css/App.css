body {
  background: #bdc3c7;
  color: white;
  padding-top: 20px;
  font-family: Roboto, sans-serif;
}

.app {
  background: white;
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  box-shadow: 0 19px 38px rgba(0, 0, 0, 0.3), 0 15px 12px rgba(0, 0, 0, 0.22);
  border-radius: 2px;
}

.app:hover {
  box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
}

.header {
  height: 60px;
  background: #3a3a3a;
  font-size: 16px;
  color: #999;
  border-radius: 2px 2px 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.display {
  height: 245px;
  padding: 20px;
  background: radial-gradient(
      123px at 50% 50%,
      rgba(34, 116, 165, 0.1) 0%,
      rgba(255, 255, 255, 0) 100%
    ),
    #1f3a93;
  background-blend-mode: lighten, normal;
}

.display__calculation {
  height: 135px;
  font-size: 20px;
  color: #cccccc;
}

.display__result {
  font-weight: 300;
  font-size: 48px;
  color: #ffffff;
  text-align: right;
}

.buttons {
  display: flex;
  flex-wrap: wrap;
}

.button {
  height: 80px;
  border-bottom: 1px solid #eee;
  border-right: 1px solid #eee;
  color: #666;
  cursor: pointer;
  /* TODO REPLACE WITH BOOSTRAP CLASS */
  display: flex;
  align-items: center;
  justify-content: center;
  transition: all 0.2s ease-out;
}

.button:hover {
  font-weight: 700;
  box-shadow: inset 0px 0px 50px 0px rgba(31, 58, 147, 0.1);
}

.button:active {
  box-shadow: inset 0px 0px 90px 0px rgba(31, 58, 147, 0.2);
}

.button--type-clear:first-child {
  background: #fb3640;
  color: white;
}

.button--type-operator {
  background: #f5f5f5;
  color: #444;
}

.button--type-equalsTo {
  background: #ddd;
  color: #222;
  border-right-color: #ddd;
  border-bottom-color: #ddd;
}

/* https://codepen.io/doggard/pen/dXYzjW */
.button:before {
  content: "";
  background-color: rgba(255, 255, 255, 0.7);
  display: block;
  position: absolute;
  border-radius: 15px;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  transform: scale(0.001, 0.001);
}

.button:active:before {
  animation: effect_dylan 0.8s ease-out;
}

@keyframes effect_dylan {
  50% {
    transform: scale(1.5, 1.5);
    opacity: 0;
  }
  99% {
    transform: scale(0.001, 0.001);
    opacity: 0;
  }
  100% {
    transform: scale(0.001, 0.001);
    opacity: 1;
  }
}
