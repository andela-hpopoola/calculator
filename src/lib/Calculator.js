import math from "mathjs";
class Calculator {
  constructor() {
    this._clearValues();
  }

  get buttons() {
    return [
      { name: "ac", label: "AC", type: "clear" },
      { name: "plusMinus", label: "+/-", type: "plusMinus" },
      { name: "delete", label: "<Del", type: "delete" },
      { name: "divide", label: "/", type: "operator" },
      { name: "seven", label: "7", type: "number" },
      { name: "eight", label: "8", type: "number" },
      { name: "nine", label: "9", type: "number" },
      { name: "multiply", label: "*", type: "operator" },
      { name: "four", label: "4", type: "number" },
      { name: "five", label: "5", type: "number" },
      { name: "six", label: "6", type: "number" },
      { name: "minus", label: "-", type: "operator" },
      { name: "one", label: "1", type: "number" },
      { name: "two", label: "2", type: "number" },
      { name: "three", label: "3", type: "number" },
      { name: "plus", label: "+", type: "operator" },
      { name: "cancel", label: "C", type: "clear" },
      { name: "decimal", label: ".", type: "decimal" },
      { name: "zero", label: "0", type: "number" },
      { name: "equals", label: "=", type: "equalsTo" }
    ];
  }

  _clearValues() {
    this.currentValue = "";
    this.display = "";
    this.result = "";
    this.lastButton = "";
  }

  _addDecimalPlace() {
    let decimal;
    if (this.currentValue.indexOf(".") < 0) {
      decimal = this.currentValue === "" ? "0." : ".";
      this.currentValue += decimal;
      this.display += decimal;
    }
  }

  _deleteLastValue() {
    if (this.display === "") {
      return;
    }

    const lastValue = this.display.endsWith(" ") ? 3 : 1;
    this.display = this.display.substring(0, this.display.length - lastValue);
    // TODO Get current value from display
    if (this.lastButton === "number" && this.currentValue !== "") {
      this.currentValue = this.currentValue.substring(
        0,
        this.currentValue.length - 1
      );
    }
  }

  _togglePlusMinusSign() {
    if (this.currentValue === "") {
      this.currentValue = "-";
      this.display += "-";
      return;
    } else if (this.currentValue === "-") {
      this.currentValue = "";
      this.display = this.display.substring(0, this.display.length - 1);
      return;
    }

    // remove currentValue from display
    this.display = this.display.substring(
      0,
      this.display.length - this.currentValue.length
    );

    // toggle sign
    this.currentValue = (parseFloat(this.currentValue) * -1).toString();

    // rewrite display
    this.display += this.currentValue;
  }

  _calculateResult() {
    try {
      this.result = String(math.eval(this.display));
    } catch (error) {
      this.result = "Error";
    }
  }

  updateCurrentValue(button) {
    if (!(button.type && button.label && button.name)) {
      return;
    }

    // ac and c button
    if (button.type === "clear") {
      this._clearValues();
    }

    // delete button is pressed
    if (button.type === "delete") {
      this._deleteLastValue();
    } else if (this.lastButton === "equalsTo") {
      // return if equalsTo button is pressed more than once
      if (button.type === "equalsTo") {
        return;
      }

      // clear display on new calculation
      this.currentValue = "";
      this.display = "";
      this.result = "";
    }

    // assign number to current value
    if (button.type === "number") {
      this.currentValue += button.label;
      this.display += button.label;
    }

    //check if plusMinus is pressed
    if (button.type === "plusMinus") {
      this._togglePlusMinusSign();
    }

    // assign decimal to current value
    if (button.type === "decimal") {
      this._addDecimalPlace();
    }

    // when operator sign is pressed
    // delete current value
    if (button.type === "operator") {
      if (this.lastButton === "operator" || this.lastButton === "decimal") {
        this._deleteLastValue();
      }

      // add zero for empty display
      if (this.display === "") {
        this.display = "0";
      }

      this.display += ` ${button.label} `;
      this.currentValue = "";
    }

    if (button.type === "equalsTo") {
      if (this.display === "") {
        this.currentValue = "";
        this.display = "";
        this.result = "0";
      } else {
        // add extra zero to create a valid expression
        const addExtraZero =
          this.lastButton === "operator" || this.lastButton === "decimal";
        if (addExtraZero) {
          this.display += "0";
        }

        // get the result
        this._calculateResult();

        this.display += ` ${button.label} `;
      }
    }

    this.lastButton = button.type;
  }

  getResult() {
    return { calculation: this.display, result: this.result };
  }
}

export default Calculator;
