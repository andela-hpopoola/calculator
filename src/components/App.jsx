import React, { Component } from "react";
import Header from "./Header";
import Display from "./Display";
import Buttons from "./Buttons";
import Calculator from "../lib/Calculator";
import "../css/App.css";
import "../css/bootstrap-grid.css";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      calculation: "",
      result: "0"
    };
    this.calculator = new Calculator();
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(button) {
    this.calculator.updateCurrentValue(button);
    this.setState(this.calculator.getResult());
  }

  render() {
    return (
      <section className="container">
        <div className="row">
          <div className="col-12 offset-0 col-sm-8 offset-sm-2 col-md-4 offset-md-4">
            <main className="app">
              <Header />
              <Display
                calculation={this.state.calculation}
                result={this.state.result}
              />
              <Buttons
                onClick={this.handleClick}
                buttonList={this.calculator.buttons}
              />
            </main>
          </div>
        </div>
      </section>
    );
  }
}

export default App;
