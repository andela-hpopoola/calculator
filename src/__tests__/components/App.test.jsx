import "../../lib/spec_helper";
import App from "../../components/App";

describe("App", () => {
  const wrapper = mount(<App />);
  const button = { name: "one", label: "1", type: "number" };

  describe("when rendered", () => {
    it("renders the Header component", () => {
      expect(wrapper.find("Header")).toHaveLength(1);
      expect(wrapper.find("Header").text()).toEqual("Calculator");
    });
    it("renders the Display component", () => {
      expect(wrapper.find("Display")).toHaveLength(1);
    });
    it("renders the Buttons component", () => {
      expect(wrapper.find("Buttons")).toHaveLength(1);
    });
  });

  it("calls click on clicked", () => {
    const instance = wrapper.instance();
    expect(wrapper.state("calculation")).toEqual("");
    instance.handleClick(button);
    expect(wrapper.state("calculation")).toEqual("1");
  });

  describe("Snapshot", () => {
    it("matches the existing snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
