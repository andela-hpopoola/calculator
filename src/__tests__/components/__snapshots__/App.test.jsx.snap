// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot matches the existing snapshot 1`] = `
<section
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-12 offset-0 col-sm-8 offset-sm-2 col-md-4 offset-md-4"
    >
      <main
        className="app"
      >
        <div
          className="header"
        >
          Calculator
        </div>
        <section
          className="display"
        >
          <header
            className="display__calculation"
          />
          <main
            className="display__result"
          >
            0
          </main>
        </section>
        <div
          className="container"
        >
          <div
            className="row"
          >
            <div
              className="button col-3 button--type-clear"
              onClick={[Function]}
            >
              AC
            </div>
            <div
              className="button col-3 button--type-plusMinus"
              onClick={[Function]}
            >
              +/-
            </div>
            <div
              className="button col-3 button--type-delete"
              onClick={[Function]}
            >
              &lt;Del
            </div>
            <div
              className="button col-3 button--type-operator"
              onClick={[Function]}
            >
              /
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              7
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              8
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              9
            </div>
            <div
              className="button col-3 button--type-operator"
              onClick={[Function]}
            >
              *
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              4
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              5
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              6
            </div>
            <div
              className="button col-3 button--type-operator"
              onClick={[Function]}
            >
              -
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              1
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              2
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              3
            </div>
            <div
              className="button col-3 button--type-operator"
              onClick={[Function]}
            >
              +
            </div>
            <div
              className="button col-3 button--type-clear"
              onClick={[Function]}
            >
              C
            </div>
            <div
              className="button col-3 button--type-decimal"
              onClick={[Function]}
            >
              .
            </div>
            <div
              className="button col-3 button--type-number"
              onClick={[Function]}
            >
              0
            </div>
            <div
              className="button col-3 button--type-equalsTo"
              onClick={[Function]}
            >
              =
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</section>
`;
